homeassistant/helpers/service.py

```diff
@@ -360,3 +362,40 @@ def async_register_admin_service(
     hass.services.async_register(
         domain, service, admin_handler, schema
     )
+
+
+@bind_hass
+@ha.callback
+def verify_domain_control(hass: HomeAssistantType, domain: str) -> Callable:
+    """Ensure permission to access any entity under domain in service call."""
+    def decorator(service: Callable) -> Callable:
+        """Decorate."""
+        async def check_permissions(call):
+            """Check user permission and raise before call if unauthorized."""
+            if not call.context.user_id:
+                return hass.async_add_job(service, call)
+
+            user = await hass.auth.async_get_user(call.context.user_id)
+            if user is None:
+                raise UnknownUser(
+                    context=call.context, permission=POLICY_CONTROL)
+
+            reg = await hass.helpers.entity_registry.async_get_registry()
+            entities = [
+                entity.entity_id for entity in reg.entities.values()
+                if entity.platform == domain
+            ]
+            _LOGGER.error(entities)
+
+            for entity_id in entities:
+                if user.permissions.check_entity(entity_id, POLICY_CONTROL):
+                    return hass.async_add_job(service, call)
+
+            raise Unauthorized(
```

Comment:
Pass in `perm_category=CAT_ENTITIES` (import constant from `auth.permissions.const`)